relays/substrate-client/src/guard.rs

```diff
@@ -0,0 +1,371 @@
+	/// Return current runtime version.
+	async fn runtime_version(&mut self) -> Result<RuntimeVersion, String>;
+	/// Return free native balance of the account on the chain.
+	async fn free_native_balance(&mut self, account: C::AccountId) -> Result<C::NativeBalance, String>;
+
+	/// Return current time.
+	fn now(&self) -> Instant {
+		Instant::now()
+	}
+	/// Sleep given amount of time.
+	async fn sleep(&mut self, duration: Duration) {
+		async_std::task::sleep(duration).await
+	}
+	/// Abort current process. Called when guard condition check fails.
+	async fn abort(&mut self) {
+		std::process::abort();
+	}
+}
+
+/// Abort when runtime spec version is different from specified.
+pub fn abort_on_spec_version_change<C: ChainWithBalances>(mut env: impl Environment<C>, expected_spec_version: u32) {
+	async_std::task::spawn(async move {
+		loop {
+			let actual_spec_version = env.runtime_version().await;
+			match actual_spec_version {
+				Ok(version) if version.spec_version == expected_spec_version => (),
+				Ok(version) => {
+					log::error!(
+						target: "bridge-guard",
+						"{} runtime spec version has changed from {} to {}. Aborting relay",
+						C::NAME,
+						expected_spec_version,
+						version.spec_version,
+					);
+
+					env.abort().await;
+				}
+				Err(error) => log::warn!(
+					target: "bridge-guard",
+					"Failed to read {} runtime version: {:?}. Relay may need to be stopped manually",
+					C::NAME,
+					error,
+				),
+			}
+
+			env.sleep(conditions_check_delay::<C>()).await;
+		}
+	});
+}
+
+/// Abort if, during a 24 hours, free balance of given account is decreased at least by given value.
+/// Other components may increase (or decrease) balance of account and it WILL affect logic of the guard.
+pub fn abort_when_account_balance_decreased<C: ChainWithBalances>(
+	mut env: impl Environment<C>,
+	account_id: C::AccountId,
+	maximal_decrease: C::NativeBalance,
+) {
+	const DAY: Duration = Duration::from_secs(60 * 60 * 24);
+
+	async_std::task::spawn(async move {
+		let mut balances = VecDeque::new();
+
+		loop {
+			let current_time = env.now();
+
+			// remember balances that are beyound 24h border
+			let time_border = current_time - DAY;
+			while balances.front().map(|(time, _)| *time < time_border).unwrap_or(false) {
+				balances.pop_front();
+			}
+
+			// read balance of the account
+			let current_balance = env.free_native_balance(account_id.clone()).await;
+
+			// remember balance and check difference
+			match current_balance {
+				Ok(current_balance) => {
+					// remember balance
+					balances.push_back((current_time, current_balance));
+
+					// check if difference between current and oldest balance is too large
+					let (oldest_time, oldest_balance) =
+						balances.front().expect("pushed to queue couple of lines above; qed");
+					let balances_difference = oldest_balance.checked_sub(&current_balance);
+					if balances_difference > Some(maximal_decrease) {
```

Comment:
Okay, it's better now, cause we compare with the balances from 24h in the past, but it won't work correctly if the account is topped up in the meantime.

Why can't we store:
```rust
last_balance: Balance;
balance_diffs: Vec<(Timestamp, bool /* sign */, Balance /* difference */)>,
```
and have a logic like this:
```rust
let diff = match current_balance.cmp(self.last_balance) {
  Ordering::Greater => Some((current_time, true, current_balance - self.last_balance)),
  Ordering::Equal => None,
  Ordering::Less => Some((current_time, false, self.last_balance - current_balance)),
};
if let Some(diff) = diff {
  self.balance_diffs.push_back(diff);
}
self.last_balance = current_balance;

let spent = self.balance_diffs.iter()
  .filter(|diff| !diff.1) // only count value drops
  .map(|diff| diff.2)
  .sum();

if spent > maximal_decrease {
 ...
}
```

It's only slightly more complex, but is more straightforward to think about.

Reply:
Well, it "won't work correctly" from your point of view. From mine, it works as it was designed. Your logic also "won't work correctly" if the account balance has been first increased by `maximal_difference` and then decreased by the same amount between two consequent checks (e.g. in the same block). We may go even further and check all events of balances module in all blocks that we see. Then the check will finally would be correct.

But I don't like where it goes. Thing that was designed to be super-simple is now turning into something that both doesn't work as you want and is checking the same thing that has been checked by the original version.

The idea of this guard is not to be super-precise. If you're ok with relayer spending 1-token/24h in regular workload, then I guess losing 2-units is acceptable. But probably relayer should be stopped in this case and you should restart it manually after making sure everything works fine. That's why there's that guard. Not to compute spent sum with absolute accuracy. Just something that is easy to deploy and to maintain. So all 3 options would work && I prefer to revert to original code to keep it simple, or just remove this guard from the relay code && rely on external components. I.e. you could just make script that is started every hour and is transferring max-tokens-to-spend-in-1-hour to relay account.

Reply:
> . Thing that was designed to be super-simple is now turning into something that both doesn't work as you want and is checking the same thing that has been checked by the original version.

I have to disagree here - afair we've been talking about alerting when we **spend** too much, not when balance drops (what was implemented). I only said that's it not equal.

> Your logic also "won't work correctly" if the account balance has been first increased by maximal_difference and then decreased by the same amount between two consequent checks (e.g. in the same block). We may go even further and check all events of balances module in all blocks that we see. Then the check will finally would be correct.

Sure, but IMHO that's edge cases. We can assume how the account is going to be used and imho it's 100% reasonable to assume that the relayer account is going to be used as follows:
1. Starts off with some initial balance
2. Spends some to perform it's work
3. At some point you realise it's reaching 0, so you top it up.

So both of your solutions are sub-optimal for the most reasonable workflow of the relayer I can think of.

> I prefer to revert to original code to keep it simple, or just remove this guard from the relay code && rely on external components.

I don't care that much any more, especially if you refuse to enable it for testnet deployments - it's basically dead code for me. Production deployment will most likely require additional steps anyway, so we may just provide a README instead.
I do like the idea of this safety valve though, but just want it to be optimal, not minimal.